File: packages/pinstripe/lib/string_reader.js

import { Base } from './base.js';

export const StringReader = Base.extend().include({
    initialize(string){
        this.string = (string || '').toString();
    },

    get length(){
        return this.string.length
    },

    toString(){
        return this.string
    },

    match(...args){
        const out = this.string.match(...args)
        if(out){
            this.string = this.string.substr(out[0].length)
        }
        return out
    }
});

File: projects/slick/lib/slick/util/html/string_reader.rb

class Slick::Util::Html::StringReader
    
    def initialize(stringable)
        @string = stringable.to_s
    end

    def length
        @string.length
    end

    def to_s
        @string
    end

    def match(pattern)
        out = @string.match(pattern)
        if out
            @string = @string[out[0].length...]
        end
        out
    end
    
end
